clustermesh/clustermesh.go

```diff
@@ -1162,3 +1214,350 @@ func (k *K8sClusterMesh) Status(ctx context.Context, log bool) (*Status, error)
+		if err != nil {
+			errs = append(errs, err.Error())
+		} else {
+			count++
+		}
+	}
+	if count > 0 {
+		k.Log("✅ Removed %d external workload resources.", count)
+	}
+	if len(errs) > 0 {
+		return errors.New(strings.Join(errs, ", "))
+	}
+	return nil
+}
+
+var installScriptFmt = `#!/bin/bash
+CILIUM_IMAGE=${1:-%[1]s}
+CLUSTER_ADDR=${2:-%[2]s}
+
+set -e
+shopt -s extglob
+
+if [ "$1" = "uninstall" ] ; then
+    if [ -n "$(docker ps -a -q -f name=cilium)" ]; then
+        echo "Shutting down running Cilium agent"
+        sudo docker rm -f cilium || true
+    fi
+    if [ -f /usr/bin/cilium ] ; then
+        echo "Removing /usr/bin/cilium"
+        sudo rm /usr/bin/cilium
+    fi
+    if [ -f /etc/resolv.conf.orig ] ; then
+        echo "Restoring /etc/resolv.conf"
+        sudo rm /etc/resolv.conf
+        sudo mv /etc/resolv.conf.orig /etc/resolv.conf
+    elif [ -f /etc/resolv.conf.link ] && [ -f $(cat /etc/resolv.conf.link) ] ; then
+        echo "Restoring systemd resolved config..."
+        if [ -f /usr/lib/systemd/resolved.conf.d/cilium-kube-dns.conf ] ; then
+	    sudo rm /usr/lib/systemd/resolved.conf.d/cilium-kube-dns.conf
+        fi
+        sudo systemctl daemon-reload
+        sudo systemctl reenable systemd-resolved.service
+        sudo service systemd-resolved restart
+        sudo rm /etc/resolv.conf
+        sudo ln -s $(cat /etc/resolv.conf.link) /etc/resolv.conf
+        sudo rm /etc/resolv.conf.link
+    fi
+    exit 0
+fi
+
+if [ -z "$CLUSTER_ADDR" ] ; then
+    echo "CLUSTER_ADDR must be defined to the IP:PORT at which the clustermesh-apiserver is reachable."
+    exit 1
+fi
+
+port='@(6553[0-5]|655[0-2][0-9]|65[0-4][0-9][0-9]|6[0-4][0-9][0-9][0-9]|[1-5][0-9][0-9][0-9][0-9]|[1-9][0-9][0-9][0-9]|[1-9][0-9][0-9]|[1-9][0-9]|[1-9])'
+byte='@(25[0-5]|2[0-4][0-9]|[1][0-9][0-9]|[1-9][0-9]|[0-9])'
+ipv4="$byte\.$byte\.$byte\.$byte"
+
+# Default port is for a HostPort service
+case "$CLUSTER_ADDR" in
+    \[+([0-9a-fA-F:])\]:$port)
+	CLUSTER_PORT=${CLUSTER_ADDR##\[*\]:}
+	CLUSTER_IP=${CLUSTER_ADDR#\[}
+	CLUSTER_IP=${CLUSTER_IP%%\]:*}
+	;;
+    [^[]$ipv4:$port)
+	CLUSTER_PORT=${CLUSTER_ADDR##*:}
+	CLUSTER_IP=${CLUSTER_ADDR%%:*}
+	;;
+    *:*)
+	echo "Malformed CLUSTER_ADDR: $CLUSTER_ADDR"
+	exit 1
+	;;
+    *)
+	CLUSTER_PORT=2379
+	CLUSTER_IP=$CLUSTER_ADDR
+	;;
+esac
+
+sudo mkdir -p /var/lib/cilium/etcd
+sudo tee /var/lib/cilium/etcd/ca.crt <<EOF >/dev/null
+%[3]sEOF
+sudo tee /var/lib/cilium/etcd/tls.crt <<EOF >/dev/null
+%[4]sEOF
+sudo tee /var/lib/cilium/etcd/tls.key <<EOF >/dev/null
+%[5]sEOF
+sudo tee /var/lib/cilium/etcd/config.yaml <<EOF >/dev/null
+---
+trusted-ca-file: /var/lib/cilium/etcd/ca.crt
+cert-file: /var/lib/cilium/etcd/tls.crt
+key-file: /var/lib/cilium/etcd/tls.key
+endpoints:
+- https://clustermesh-apiserver.cilium.io:$CLUSTER_PORT
+EOF
+
+CILIUM_OPTS=" --join-cluster --enable-host-reachable-services --enable-endpoint-health-checking=false"
```

Comment:
Host reachable services require at least 4.17 kernel as far as I can tell. Are they strictly needed or should we document the fact that external workloads require at least kernel 4.17?

Reply:
Yes, strictly needed and currently documented on the GSG. Maybe adding a kernel version check to the generated installation script would be good. Would parsing `uname -r` output be portable enough?

```
$ uname -r
4.19.177-0419177-generic
```

Reply:
Indeed, sorry. I'm not sure how I have missed this in the GSG :facepalm:

Agree, I think for this case a check of `uname -r` checking against the versions mentioned in https://docs.cilium.io/en/v1.9/gettingstarted/host-services/ should probably be portable enough.

Reply:
The GSG now states the kernel version explicitly and links to the host reachable service docs. For me this is sufficent for now. We can work on a kernel version check in a follow-up PR.